terminal/tcell/tcell.go

```diff
@@ -0,0 +1,205 @@
+	"github.com/gdamore/tcell/encoding"
+	"github.com/mum4k/termdash/cell"
+	"github.com/mum4k/termdash/internal/event/eventqueue"
+	"github.com/mum4k/termdash/terminal/terminalapi"
+)
+
+// Option is used to provide options.
+type Option interface {
+	// set sets the provided option.
+	set(*Terminal)
+}
+
+// option implements Option.
+type option func(*Terminal)
+
+// set implements Option.set.
+func (o option) set(t *Terminal) {
+	o(t)
+}
+
+// DefaultColorMode is the default value for the ColorMode option.
+const DefaultColorMode = terminalapi.ColorMode256
+
+// ColorMode sets the terminal color mode.
+// Defaults to DefaultColorMode.
+func ColorMode(cm terminalapi.ColorMode) Option {
+	return option(func(t *Terminal) {
+		t.colorMode = cm
+	})
+}
+
+// ClearStyle sets the style to use for tcell when clearing the screen.
+// Defaults to white foreground and black background.
+func ClearStyle(fg, bg cell.Color) Option {
+	return option(func(t *Terminal) {
+		t.clearStyle = &cell.Options{
+			FgColor: fg,
+			BgColor: bg,
+		}
+	})
+}
+
+// Terminal provides input and output to a real terminal. Wraps the
+// gdamore/tcell terminal implementation. This object is not thread-safe.
+// Implements terminalapi.Terminal.
+type Terminal struct {
+	// events is a queue of input events.
+	events *eventqueue.Unbound
+
+	// done gets closed when Close() is called.
+	done chan struct{}
+
+	// the tcell terminal window
+	screen tcell.Screen
+
+	// Options.
+	colorMode  terminalapi.ColorMode
+	clearStyle *cell.Options
+}
+
+// newTerminal creates the terminal and applies the options.
+func newTerminal(opts ...Option) (*Terminal, error) {
+	screen, err := tcell.NewScreen()
+	if err != nil {
+		return nil, err
+	}
+
+	t := &Terminal{
+		events:    eventqueue.New(),
+		done:      make(chan struct{}),
+		colorMode: DefaultColorMode,
+		clearStyle: &cell.Options{
+			FgColor: cell.ColorWhite,
+			BgColor: cell.ColorBlack,
+		},
+		screen: screen,
+	}
+	for _, opt := range opts {
+		opt.set(t)
+	}
+
+	return t, nil
+}
+
+// New returns a new tcell based Terminal.
+// Call Close() when the terminal isn't required anymore.
+func New(opts ...Option) (*Terminal, error) {
+	// Enable full character set support for tcell
+	encoding.Register()
+
+	t, err := newTerminal(opts...)
+	if err != nil {
+		return nil, err
+	}
+	if err = t.screen.Init(); err != nil {
+		return nil, err
+	}
+
+	clearStyle := cellOptsToStyle(t.clearStyle, t.colorMode)
+	t.screen.EnableMouse()
+	t.screen.SetStyle(clearStyle)
+
+	go t.pollEvents() // Stops when Close() is called.
+	return t, nil
+}
+
+// Size implements terminalapi.Terminal.Size.
+func (t *Terminal) Size() image.Point {
+	w, h := t.screen.Size()
+	return image.Point{
+		X: w,
+		Y: h,
+	}
+}
+
+// Clear implements terminalapi.Terminal.Clear.
+func (t *Terminal) Clear(opts ...cell.Option) error {
+	o := cell.NewOptions(opts...)
+	st := cellOptsToStyle(o, t.colorMode)
+	w, h := t.screen.Size()
+	for row := 0; row < h; row++ {
```

Comment:
Can we use the Fill() method on the screen object instead or do we need to iterate ourselves?

https://github.com/gdamore/tcell/blob/b03a4607a44a5d8b4d4a0b9d2fde3ced0c28e78c/screen.go#L33

Reply:
:+1: Will look into it, originally used the code from tcell's termbox compat file which manually iterated and set the content to ' '.

Reply:
Switched to Fill() and it works fine.